MaxText/sft/hooks.py

```diff
@@ -0,0 +1,182 @@
+      self.metric_logger.flush_metrics_and_cleanup()
+
+  @override
+  def on_train_step_start(self, train_ctx: peft_trainer.PeftTrainer):
+    """Called at the beginning of a training step."""
+    self.train_metadata["step_start_time"] = datetime.datetime.now()
+
+    if self.config.enable_goodput_recording:
+      record_goodput(self.goodput_recorder, f"record_{GoodputEvent.STEP.value}_start_time", train_ctx.train_steps)
+
+    # Calculate the number of non-padded tokens in the batch
+    self.train_metadata["total_weights"] = jnp.sum(train_ctx.data_hooks.train_batch["targets_segmentation"] != 0)
+
+  @override
+  def on_train_step_end(self, train_ctx: peft_trainer.PeftTrainer, train_loss: float):
+    """Called at the end of training step."""
+    assert (
+        self.train_metadata["step_start_time"] != 0
+    ), "SFTTrainingHooks.on_train_step_start() must be called before SFTTrainingHooks.on_train_step_end()"
+
+    step_time_delta = datetime.datetime.now() - self.train_metadata["step_start_time"]
+
+    # We check against `train_ctx.train_steps - 1` because this hook is called
+    # after `train_ctx.train_steps` has been incremented for the current
+    # step.
+    if self.metadata["first_train_step"] == train_ctx.train_steps - 1:
+      max_utils.print_mem_stats("After params initialized")
+
+    metrics = {
+        "scalar": {
+            "learning/loss": train_loss,
+            "learning/total_weights": self.train_metadata["total_weights"],
+        }
+    }
+    self.metric_logger.record_train_metrics(metrics, train_ctx.train_steps - 1, step_time_delta)
+    self.metric_logger.write_metrics(metrics, train_ctx.train_steps - 1)
+    self.train_metadata.clear()
+
+  @override
+  def on_eval_step_start(self, train_ctx: peft_trainer.PeftTrainer):
+    """Called at the beginning of an evaluation step."""
+    self.eval_metadata["eval_step_count"] += 1.0
+    # Calculate the number of non-padded tokens in the batch
+    self.eval_metadata["total_weights"] += jnp.sum(train_ctx.data_hooks.eval_batch["targets_segmentation"] != 0)
+
+  @override
+  def on_eval_step_end(self, train_ctx: peft_trainer.PeftTrainer, eval_loss: float):
+    """Called at the end of evaluation step."""
+    assert (
+        self.eval_metadata["eval_step_count"] != 0
+    ), "SFTTrainingHooks.on_eval_step_start() must be called before SFTTrainingHooks.on_eval_step_end()"
+
+    avg_loss = eval_loss / self.eval_metadata["eval_step_count"]
+    metrics = {
+        "scalar": {
+            "eval/total_loss": eval_loss,
+            "eval/avg_loss": avg_loss,
+            "eval/total_weights": self.eval_metadata["total_weights"],
+        }
+    }
+    self.metric_logger.write_metrics(metrics, train_ctx.train_steps, is_training=False)
+    self.eval_metadata.clear()
+
+    if avg_loss <= self.config.target_eval_loss:
+      raise exceptions.StopTraining(f"Target loss {self.config.target_eval_loss=} is achieved.")
+
+
+class SFTDataHooks(DataHooks):
```

Comment:
Do we need DataHooks?

Instead, Tunix could consume a DataSource (from MaxText or any other source) with a minimal, checkpointable interface:
`next_train(), next_eval(), state(), load_state(), reset(), etc`.

This keeps data lifecycle (prefetch, retries) in one place, makes it checkpointable, and moves eval cadence/early stop into the trainer.
I would simplify swapping in other data sources and remove coupling between hooks and data loading.